ipn/ipnauth/actor.go

```diff
@@ -45,3 +48,51 @@ type ActorCloser interface {
 	// Close releases resources associated with the receiver.
 	Close() error
 }
+
+// SessionID is an opaque, comparable value used to identify a logical [Session].
+type SessionID struct {
+	v any
+}
+
+type uniqueID struct {
+	p *byte
+	s string
+}
+
+// NewSessionID returns a new process-unique [SessionID].
+func NewSessionID() SessionID {
+	// We use both a byte pointer and a random hex string out of paranoia that
+	// eventually we may use a SessionID in a context where 64 bits of entropy
+	// of the hex string is not enough. The string is nice for debugging and logging,
+	// but the pointer guarantees uniqueness of the ID within the address space
+	// of a given process.
+	return SessionID{uniqueID{new(byte), rands.HexString(16)}}
+}
+
+// SessionIDFrom returns a new [SessionID] derived from the specified value.
+// SessionIDs derived from equal values are equal.
+func SessionIDFrom[T comparable](v T) SessionID {
+	return SessionID{v}
+}
+
+// String implements [fmt.Stringer].
+func (id SessionID) String() string {
+	if id.v == nil {
+		return "(none)"
+	}
+	return fmt.Sprint(id.v)
+}
+
+func (id uniqueID) String() string {
+	return id.s
+}
+
+// Session is an optional interface that an [Actor] might implement to link a
+// series of related interactions with the LocalAPI by a [SessionID].
```

Comment:
I'm not a fan of this. Both the optional interface, and it hanging off an Actor of all things. That's not the relationship direction that makes sense to me.

And that's aside from us already having a session concept and adding a new one.

Reply:
The current session concept is built on top of and around the notification watcher sessions. This results in a few key points:
- a "session" only exists if a client starts watching notifications;
- at the same time, it's confusing that a single client can open more than one `watchSession`. There's really no 1:1 relationship between clients/logical sessions and these `watchSession`s;
- therefore, it does not allow associating user actions (e.g., a user calling `StartLoginInteractive`) with a specific `watchSession`. This means we can't send notifications exclusively to the client that initiated a given action. For example, if the same user is utilizing both the GUI and the CLI, we may want to avoid sending the `authURL` to the GUI when the user is interacting via the CLI;
- I don't particularly like the current approach from a security perspective. This doesn't mean the proposed changes is the only alternative; for example, we could authenticate clients and pass session tokens around, maintaining a `map[tokenId]ipnauth.Actor`. However, I don't want to change the APIs too much or introduce an authentication flow before clients can essentially do `tailscale up`.

Re: "the optional interface, and it hanging off an Actor of all things."
LocalAPI is not the only way to use the `LocalBackend`. We use its exported methods directly in several places, including `tsnet`, `tailscaled`, and others. The concept of a "session" is not universal; it's optional. There's no "session" when we call `(*LocalBackend).Start()` from `tsnet` or `tailscaled`, or when the Android client calls `(*LocalBackend).WatchNotifications()`. However, there's always an actor who initiated that action, even if it's `tailscaled` itself, which should likely have unrestricted access to the `LocalBackend`.

Or, to put it another way, whenever the `LocalBackend` is used, there's always an actor involved. This could be a local user, a remote user, or the daemon itself. In some cases, we can even narrow the actor down to a specific session opened by that user (e.g., a specific local process or an H2C connection), while in other cases, it's not possible.

Reply:
I think this all calls for a design doc/discussion on the whole concept of sessions. I'd rather not do the design discussion on misc PRs (this one and the h2c one and whatever other ones are coming). Let's figure out the overall plan first.

Reply:
I believe this is now resolved after discussing it on the call and making the requested changes.